File: grunt/countries.js


import _ from 'lodash';
import * as utils from '../src/javascripts/utils'


export default (grunt) => {

  grunt.registerTask(
    'countries-data',
    'Filter country labels GeoJSON',
    () => {

      let labels = [];

      // Load OSM JSON.
      let json = grunt.file.readJSON('data/labels.geo.json');

      // Walk countries.
      for (let c of json.features) {

        let name = c.properties['name:en'];

        // lon/lat -> XYZ.
        let [x, y, z] = utils.lonLatToXYZ(
          c.geometry.coordinates[0],
          c.geometry.coordinates[1]
        );

        if (name) {
          labels.push({
            name: name,
            x: x,
            y: y,
            z: z,
          });
        }

      }

      grunt.file.write(
        'src/javascripts/data/labels.json',
        JSON.stringify(labels, null, 2)
      );

    }
  );

  grunt.registerTask(
    'countries-sprite',
    'Render country labels sprite sheet',
    () => {

      console.log('sprite');

    }
  );

};

File: tasks/countries.py


import json

from invoke import task


@task
def merge_labels():

    """
    Merge label names.
    """

    countries = open_countries()

    labels = cca3_to_label()
    for c in countries['features']:
        c['properties']['label'] = labels.get(c['id'])

    write_countries(countries)


@task
def merge_areas():

    """
    Merge country areas.
    """

    countries = open_countries()

    areas = cca3_to_area()
    for c in countries['features']:
        c['properties']['area'] = areas.get(c['id'])

    write_countries(countries)


@task
def merge_anchors():

    """
    Merge label anchor points.
    """

    countries = open_countries()

    anchors = cca3_to_anchor()
    for c in countries['features']:
        c['properties']['anchor'] = anchors.get(c['id'])

    write_countries(countries)


@task(merge_labels, merge_anchors, merge_areas)
def merge():
    pass


def open_countries():

    """
    Read the borders GeoJSON.

    Returns: dict
    """

    with open('src/javascripts/data/countries.json', 'r') as fh:
        return json.loads(fh.read())


def write_countries(countries):

    """
    Write updated GeoJSON.

    Args:
        countries (dict)
    """

    with open('src/javascripts/data/countries.json', 'w') as fh:
        json.dump(countries, fh, sort_keys=True)


def open_mledoze():

    """
    Read the meldoze metadata.

    Returns: dict
    """

    with open('data/countries/mledoze.json', 'r') as fh:
        return json.loads(fh.read())


def cca3_to_area():

    """
    Map country code -> area.

    Returns: dict
    """

    areas = {}
    for c in open_mledoze():
        areas[c['cca3']] = c['area']

    return areas


def cca3_to_anchor():

    """
    Map country code -> anchor point.

    Returns: dict
    """

    anchors = {}
    for c in open_mledoze():

        # Swap to (lon, lat).
        a = c['latlng']
        if a: anchors[c['cca3']] = [a[1], a[0]]

    return anchors


def cca3_to_label():

    """
    Map country code -> common name.

    Returns: dict
    """

    labels = {}
    for c in open_mledoze():
        labels[c['cca3']] = c['name']['common']

    return labels
